### PR TITLE
Add Indices of X in List

### PR DESCRIPTION
**Pull request by Burbulinis**
### Description
This PR aims to add an expression to get the indices or positions of a list where the value at that index is the provided value.
The indices option is only allowed for variable lists, which returns strings. Whereas for positions it returns an integer. You can get the position of any list. For example:

```applescript
set {_list::a} to 1
set {_list::b} to 2
set {_list::c} to 1
set {_list::d} to 2
set {_list::e} to 1

send all indices of the value 2 in {_list::*} # returns "b" and "d"
send all positions of the value 1 in {_list::*} # returns 1, 3, 5

send positions of value {_whatever} in (some expression that returns multiple values)
# allows non-literals and non-variable lists

# also allows for strings:

send all positions of "a" in "abcabcabc" # returns 1, 4 and 7

send first/last position of "breh" in "whatever breh" 
# added support for the keyword 'position' and support to get all of the positions in the string
```

This expression was merged with the current [ExprIndexOf](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/expressions/ExprIndexOf.java), which I renamed for clarity. The default, if you do not specify last or all is the first index/position. This is done to be consistent with the old ExprIndexOf.

**BREAKING CHANGE**:
- returning nothing instead of `-1` for an unfound value in a String. (i.e. `index of "a" in "b"`)

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** https://github.com/SkriptLang/Skript/issues/3933

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	private enum IndexType {
+		FIRST, LAST, ALL;
+
+		public static IndexType get(int mark) {
+			return switch (mark) {
+				case 1 -> FIRST;
+				case 2 -> LAST;
+				default -> ALL;
+			};
+		}
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:133]: can be replaced with `IndexType.values()[mark]`

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+		@Override
+		public String toString() {
+			return name().toLowerCase();
+		}
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:138]: this can be replaced with `type.name().toLowerCase()`

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@Name("Indices of X in List")
+@Description("Returns all the indices of a list where their value is X.")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:24]:
```suggestion
@Description("Returns all the indices of a list where the value at that index is the provided value.")
```

**Review by Efnilite — CHANGES_REQUESTED** [review]
very nois

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	private IndexType type;
+	private Expression<?> value;
+	private Variable<?> list;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (!(exprs[1] instanceof Variable<?> var) || !var.isList()) {
+			Skript.error("'" + exprs[1].toString(null, false) +
+				"' can only ever have one value at most, thus the 'indices of x in ...' expression is useless.");
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:57]:
```suggestion
				"' can only ever have one value at most, thus the 'indices of x in list' expression has no effect.");
```

**Review by Efnilite — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java (hunk 4 above)
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:57]: Is this true? Can't a user pass in an expression here which returns multiple values which isn't a variable?

Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:57]: it checks if it's a variable & list beforehand and I think it should only work with variables.

Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java]: you are indeed right

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	"",
+	"set {_otherlist::burb} to 100",
+	"set {_otherlist::burp} to 100",
+	"set {_indices::*} to the first index of the value 100 in {_otherlist::*}",
+	"# {_indices::*} is now \"burb\"",
+	"set {_indices::*} to the last index of the value 100 in {_otherlist::*}",
+	"# {_indices::*} is now \"burp\""
+})
+@Since("INSERT VERSION")
+public class ExprIndicesOfX extends SimpleExpression<String> {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:41]: Thoughts about returning numbers for non-variable lists?
```
index of player in all players -> 4
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+	static {
+		Skript.registerExpression(ExprIndicesOfX.class, String.class, ExpressionType.COMBINED,
+			"[the] [1:first|2:last] (indices|index[es]) of [[the] value] %object% in %objects%"
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:45]:
```suggestion
			"[the] [1:first|2:last] (indices|index[es]) of [[the] value] %object% in %~objects%"
```
(only if sticking to variables only)

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	private IndexType type;
+	private Expression<?> value;
+	private Variable<?> list;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (!(exprs[1] instanceof Variable<?> var)) {
+			Skript.error("'" + exprs[1].toString(null, false) + "' is not a list variable.");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:56]:
```suggestion
			Skript.error("'" + exprs[1].toString(null, false) + "' is not a list variable. Only list variables may be searched for indices.");
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		Map<String, Object> variable = (Map<String, Object>) list.getRaw(event);
+		if (variable == null)
+			return new String[0];
+
+		List<String> indices = new ArrayList<>();
+
+		for (Map.Entry<String, Object> entry : variable.entrySet()) {
+			Object entryValue = entry.getValue();
+			if (entryValue instanceof Map<?, ?> map)
+				entryValue = map.get(null);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:88]: why are you checking this? this would include `{var::value}` in the list `{var::value::*}`

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java (hunk 8 above)
Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:88]: huh, should it not be like this? e.g.
```
{list::1} -> 10
{list::1::1} -> 10
{list::2} -> 10
```
and do `indices of value 10 in {list::*}` would return `1` and `2`
if i were to not do that, it would only return `2`

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:88]: oh i see, I was thinking it was acting on the map one level up

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+		"",
+		"set {_positions::*} to the positions of the value 100 in {_otherlist::*}",
+		"# {_positions::*} is now 1, 2, 3",
+})
+@Since("INSERT VERSION")
+public class ExprIndicesOfX extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprIndicesOfX.class, Object.class, ExpressionType.COMBINED,
+			"[the] [1:first|2:last] (indices|index[es]|:position[s]) of [[the] value] %object% in %objects%"
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:52]: multiple patterns please

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	private IndexType type;
+	private boolean position;
+	private Expression<?> value;
+	private Expression<?> objects;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (exprs[1].isSingle() || (exprs[1] instanceof Variable<?> var && !var.isList())) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:63]: the first check covers the second, so the second is redundant

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@Name("Indices of X in List")
+@Description(
+		"Returns the indices or positions of a list where the value at that index is the provided value. " +
+		"Getting the indices of a list will return strings, whereas the positions will return a number."
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:23]:
```suggestion
		"Returns the indices or positions of a list where the value at that index is the provided value. " +
		"Indices are only supported for variable lists and will return the string indices of the given value. " +
		"Positions can be used with any list and will return the numerical position of the value in the list, counting up from 1."
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+				}
+				count++;
+			}
+		} else {
+			for (Object object : objects.getArray(event)) {
+				if (object.equals(value)) {
+					if (position)
+						indices.add(count);
+					else
+						indices.add(String.valueOf(count));
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:111]: I don't think we should support string indices on non-variables.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java (hunk 10 above)
Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:63]: ooo that's cool, i see

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+					indices.add(index);
+				}
+				position++;
+			}
+		} else {
+			for (Object object : objects.getArray(event)) {
+				if (object.equals(value)) {
+					if (type == IndexType.FIRST)
+						return new Object[]{position};
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:128]: You're saying the return type is Integer or String, but you're returning Object[] here.

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+				Object entryValue = entry.getValue();
+				// the value of {foo::1} when {foo::1::bar} is set is a map with a null key of the value {foo::1}
+				if (entryValue instanceof Map<?, ?> map)
+					entryValue = map.get(null);
+
+				if (entryValue.equals(value)) {
+					Object index = getPositionOrIndex(entry.getKey(), position, this.position);
+
+					if (type == IndexType.FIRST)
+						return new Object[]{index};
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:118]: Also here.

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+				}
+				position++;
+			}
+		}
+
+		if (indices.isEmpty())
+			return new Object[0];
+
+		if (type == IndexType.LAST)
+			return new Object[]{indices.get(indices.size() - 1)};
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:140]: And here.

**Review by Moderocky — CHANGES_REQUESTED** [review]

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+		if (this.position)
+			return indices.toArray(new Integer[0]);
+		return indices.toArray(new String[0]);
+	}
+
+	private Object getPositionOrIndex(String key, int position, boolean count) {
+		if (count)
+			return position;
+		return key;
+	}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:165]: this doesn't need to exist, it's only ever used once and doesn't simplify anything, really.

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+					indices.add(position);
+				}
+				position++;
+			}
+		}
+
+		if (indices.isEmpty()) {
+			if (this.position)
+				return new Integer[0];
+			return new String[0];
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:147]: all these int or string arrays could be replaced with either a helper method or a call to Array.newInstance with a stored class (determined in init)

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		SyntaxStringBuilder builder = new SyntaxStringBuilder(event, debug);
+
+		builder.append(type.name().toLowerCase());
+		if (type == IndexType.ALL) {
+			builder.append("indices");
+		} else if (position) {
+			builder.append("positions");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:187]: this logic isn't correct. all positions would print indices

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+	private IndexType type;
+	private boolean position;
+	private Expression<?> value;
+	private Expression<?> objects;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (exprs[1].isSingle()) {
+			Skript.error("'" + exprs[1].toString(null, false) +
+					"' can only ever have one value at most, thus the 'indices of x in list' expression has no effect.");
```
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java]: toString(null, false) is the established precedent for this error (see ExprAmount)

[the remark this answers is not in the document]
Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java]: `expression.toString()` just calls `expression.toString(null, false)`

[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java]: oh cool didn't know that

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+		"set {_positions::*} to the positions of the value 100 in {_otherlist::*}",
+		"# {_positions::*} is now 1, 2, 3",
+})
+@Since("INSERT VERSION")
+public class ExprIndicesOfX extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprIndicesOfX.class, Object.class, ExpressionType.COMBINED,
+			"[the] [1:first|2:last] (indices|index[es]) of [[the] value] %object% in %objects%",
+			"[the] [1:first|2:last] position[s] of [[the] value] %object% in %objects%"
```
[the remark this answers is not in the document]
Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:62]: In indices, sure, but for positions I allow literals to get the position of the value, e.g.
`set {_positions::*} to positions of value x in (some expression that returns multiple x)`

[the remark this answers is not in the document]
Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:62]: yeah that's a non-literal. Literals lists are like `"foo", "bar" and "foobar"`

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+				}
+				position++;
+			}
+
+			if (type == IndexType.LAST)
+				return new Integer[]{positions.get(positions.size() - 1)};
+			return positions.toArray();
+		}
+
+		Variable<?> list = (Variable<?>) objects;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:113]: Unsafe, need to check if it's actually a variable

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — CHANGES_REQUESTED** [review]
I have forgotten that ExprIndexOf already exists. This needs to be merged with that

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java (hunk 21 above)
Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:113]: It can never be not a variable. Positions is always true when it's a non-variable

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java:113]: Oh I see you've got a check up in init
Just throw an assert here then

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+		string = matchedPattern == 2;
+		value = LiteralUtils.defendExpression(exprs[0]);
+		objects = exprs[1];
+
+		return LiteralUtils.canInitSafely(value);
+	}
+	
+	@Override
+	@Nullable
+	protected Object[] get(Event event) {
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:107]:
```suggestion
	protected Object @Nullable [] get(Event event) {
```

**Review by TheAbsolutionism — COMMENTED** [review]
My very important review

**Review by TheAbsolutionism — APPROVED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprIndicesOfX.sk [hunk 23]
```diff
@@ -0,0 +1,28 @@
+test "index of":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprIndicesOfX.sk:1]: needs tests for strings

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+	"# {_positions::*} is now 2 and 4"
+})
+@Since("2.1, INSERT VERSION (indices, positions of list)")
+public class ExprIndicesOf extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprIndicesOf.class, Object.class, ExpressionType.COMBINED,
+			"[the] [first|1:last|2:all] (indices|index[es]) of [[the] value] %object% in %~objects%",
+			"[the] [first|1:last|2:all] position[s] of [[the] value] %object% in %~objects%",
+			"[the] [first|1:last|2:all] (position[s]|indices|index[es]) of [[the] value] %string% in %string%"
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:75]: won't this be overshadowed by the first 2 patterns
it probably needs to be first

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+
+/**
+ * @author Peter Güttinger
+ */
+@Name("Index Of")
+@Description({
+	"The first, last or all indices of a character (or text) in a text, or -1 if it doesn't occur in the text."
+		+ "Indices range from 1 to the <a href='#ExprIndicesOf'>length</a> of the text.",
+	"You can also get the indices or positions of a list where the value at that index is the provided value. "
+		+ "Indices are only supported for variable lists and will return the string indices of the given value. "
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:34]: maybe reword this to make it clear that this doesn't apply to string on string violence

**Review by sovdeeth — APPROVED** [review]
code looks good

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+})
+@Since("2.1, INSERT VERSION (indices, positions of list)")
+public class ExprIndicesOf extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprIndicesOf.class, Object.class, ExpressionType.COMBINED,
+			"[the] [first|1:last|2:all] (position[s]|indices|index[es]) of [[the] value] %string% in %string%",
+
+			"[the] [first|1:last|2:all] (position[s]|indices|index[es]) of [[the] value] %string%"
+				+ "in %string% from [[the] position] %integer%",
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:89]: do you think this complexity is worth it compared to just making the user use the `substring` expression? Likewise for lists with the element of expression

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:89]: I also don't like how you're using `position` to mean two different things in the same syntax

Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:89]: what would be the ideal replacement? just simply `from x`?

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:89]: Well substring makes it clear, but you could use `from character x`

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * @author Peter Güttinger
+ */
+@Name("Indices Of")
+@Description({
+	"Get the first, last or all positions of a character (or text) in another text using "
+		+ "'positions of %text% in %text%'. -1 is returned when the value does not occur in the text."
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:32]:
```suggestion
		+ "'positions of %text% in %text%'. -1 is returned when the value does not occur in the text. "
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+		+ "Positions range from 1 to the <a href='#ExprIndicesOf'>length</a> of the text.",
+	"",
+	"Using 'indices/positions of %objects%', you can get the indices or positions of a list where the value at that index is the provided value. "
+		+ "Indices are only supported for variable lists and will return the string indices of the given value. "
+		+ "Positions can be used with any list and will return the numerical position of the value in the list, counting up from 1."
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:37]:
```suggestion
		+ "Positions can be used with any list and will return the numerical position of the value in the list, counting up from 1. "
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 29]
```diff
@@ -0,0 +1,246 @@
...
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * @author Peter Güttinger
+ */
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:28]:
```suggestion
```

**Review by Efnilite — APPROVED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprIndicesOfX.sk
Comment (Efnilite) [inline src/test/skript/tests/syntaxes/expressions/ExprIndicesOfX.sk:1]: file name is different frmo the expr

Comment (Burbulinis) [reply src/test/skript/tests/syntaxes/expressions/ExprIndicesOfX.sk:1]: oops

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 30]
```diff
@@ -0,0 +1,243 @@
...
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+@Name("Indices Of")
+@Description({
+	"Get the first, last or all positions of a character (or text) in another text using "
+		+ "'positions of %text% in %text%'. -1 is returned when the value does not occur in the text. "
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:29]: This should return nothing to match the behaviour of `positions of %objects%`

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 31]
```diff
@@ -0,0 +1,243 @@
...
+		+ "Positions range from 1 to the <a href='#ExprIndicesOf'>length</a> of the text.",
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:30]:
```suggestion
		+ "Positions range from 1 to the <a href='#ExprIndicesOf'>length</a> of the text (inclusive).",
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 32]
```diff
@@ -0,0 +1,243 @@
...
+	"",
+	"Using 'indices/positions of %objects%', you can get the indices or positions of a list where the value at that index is the provided value. "
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:32]:
```suggestion
	"Using 'indices/positions of %objects% in %objects%', you can get the indices or positions of a list where the value at that index is the provided value. "
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 33]
```diff
@@ -0,0 +1,243 @@
...
+			if (string) {
+				String haystack = (String) objects.getSingle(event);
+				if (haystack == null) {
+					return new Long[0];
+				}
+
+				return getStringPositions(haystack, (String) value);
+			}
+
+			Object[] list = objects.getArray(event);
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:123]: You should use iterators (`Expression#iterator`) to avoid unnecessary computations

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 34]
```diff
@@ -0,0 +1,243 @@
...
+			}
+
+			return getListPositions(list, value);
+		}
+
+		assert objects instanceof Variable<?>;
+
+		Variable<?> list = (Variable<?>) objects;
+		//noinspection unchecked
+		Map<String, Object> variable = (Map<String, Object>) list.getRaw(event);
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:135]: Same as above (`Variable#variablesIterator`)

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 35]
```diff
@@ -0,0 +1,243 @@
...
+				continue;
+			}
+
+			if (type == IndexType.FIRST) {
+				return new Long[]{position};
+			}
+			positions.add(position);
+
+			position++;
+		}
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:178]:
```suggestion
		long position = 0;
		for (Object object : list) {
			position++;
			if (!object.equals(value))
				continue;

			if (type == IndexType.FIRST) {
				return new Long[]{position};
			}
			positions.add(position);
		}
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 36]
```diff
@@ -0,0 +1,243 @@
...
+
+		return positions.toArray(new Long[0]);
+	}
+
+	private String[] getVariableIndices(Map<String, Object> variable, Object value) {
+		List<String> indices = new ArrayList<>();
+
+		for (Entry<String, Object> entry : variable.entrySet()) {
+			Object entryValue = entry.getValue();
+			// the value of {foo::1} when {foo::1::bar} is set is a map with a null key of the value {foo::1}
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:192]: is set is a map?

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java (hunk 30 above)
Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:30]: that would become a breaking change. I talked to sovde about this and he said I should keep it as is
https://discord.com/channels/135877399391764480/836220422223036467/1323675600707260426

Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:27]: [have changed it](https://discord.com/channels/135877399391764480/836220422223036467/1341859111943405598)

**Review by sovdeeth — APPROVED** [review]

**Review by UnderscoreTud — APPROVED** [review]

**Review by sovdeeth — CHANGES_REQUESTED** [review]
re request when tests fixed

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 37]
```diff
@@ -0,0 +1,243 @@
...
+	"# {_positions::*} is now 2 and 4"
+})
+@Since("2.1, INSERT VERSION (indices, positions of list)")
+public class ExprIndicesOf extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprIndicesOf.class, Object.class, ExpressionType.COMBINED,
+			"[the] [1:first|2:last|3:all] (position[mult:s]|mult:indices|index[mult:es]) of [[the] value] %string% in %string%",
+			"[the] [1:first|2:last|3:all] position[mult:s] of [[the] value] %object% in %~objects%",
+			"[the] [1:first|2:last|3:all] (mult:indices|index[mult:es]) of [[the] value] %object% in %~objects%"
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:76]:
```suggestion
			"[the] [first|1:last|2:all] (position[mult:s]|mult:indices|index[mult:es]) of [[the] value] %string% in %string%",
			"[the] [first|1:last|2:all] position[mult:s] of [[the] value] %object% in %~objects%",
			"[the] [first|1:last|2:all] (mult:indices|index[mult:es]) of [[the] value] %object% in %~objects%"
```

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java [hunk 38]
```diff
@@ -0,0 +1,243 @@
...
+			return false;
+		}
+
+		if (!(exprs[1] instanceof Variable<?>) && matchedPattern == 2) {
+			Skript.error("'" + exprs[1] + "' is not a list variable. "
+				+ "You can only get the indices of a list variable.");
+			return false;
+		}
+
+		indexType = IndexType.values()[parseResult.mark == 0 ? 0 : parseResult.mark - 1];
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:98]:
```suggestion
		indexType = IndexType.values()[parseResult.mark];
```

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java (hunk 37 above)
Comment (Burbulinis) [reply src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java:76]: that would mean if i were to do `first positions of the value ...`, it would return them all
it should only return them all if i dont specify first/last/all and use the plural form

**Review by sovdeeth — APPROVED** [review]
